File: internal/platform/monitoring/service/monitoring.go
package service

import (
	"context"
	"fmt"
	"time"

	"github.com/google/uuid"

	"Logos/internal/platform/monitoring/dao"
	"Logos/internal/platform/monitoring/model"
)

type MonitoringService interface {
	RecordMetric(ctx context.Context, serviceName string, metricType int32, value float64, unit string, tags map[string]string) error
	BatchRecordMetrics(ctx context.Context, metrics []struct {
		ServiceName string
		Type        int32
		Value       float64
		Unit        string
		Tags        map[string]string
	}) error
	QueryMetrics(ctx context.Context, serviceName string, metricType int32, startTime, endTime int64, tags map[string]string) ([]*model.Metric, int64, error)

	RecordLog(ctx context.Context, serviceName, level, message string, fields map[string]string) error
	BatchRecordLogs(ctx context.Context, logs []struct {
		ServiceName string
		Level       string
		Message     string
		Fields      map[string]string
	}) error
	QueryLogs(ctx context.Context, serviceName, level, query string, startTime, endTime int64, page, pageSize int32) ([]*model.LogEntry, int64, error)

	QueryAlerts(ctx context.Context, serviceName *string, level *int32, resolved *bool, startTime, endTime int64) ([]*model.Alert, int64, error)
	ResolveAlert(ctx context.Context, alertID string) error

	UpdateServiceStatus(ctx context.Context, serviceName, status string, errorMessage *string, metadata map[string]string) error
	GetServiceStatus(ctx context.Context, serviceName string) (*model.ServiceStatus, error)
	ListServiceStatuses(ctx context.Context) ([]*model.ServiceStatus, error)
}

type monitoringServiceImpl struct {
	repo dao.MonitoringRepository
}

func NewMonitoringService(repo dao.MonitoringRepository) MonitoringService {
	return &monitoringServiceImpl{repo: repo}
}

func (s *monitoringServiceImpl) RecordMetric(ctx context.Context, serviceName string, metricType int32, value float64, unit string, tags map[string]string) error {
	m := &model.Metric{
		ID:          uuid.New().String(),
		ServiceName: serviceName,
		Type:        int(metricType),
		Value:       value,
		Unit:        unit,
		Tags:        dao.MarshalTags(tags),
		Timestamp:   time.Now().UnixMilli(),
	}

	if err := s.repo.SaveMetric(ctx, m); err != nil {
		return fmt.Errorf("保存指标失败: %w", err)
	}
	s.checkThreshold(ctx, m)
	return nil
}

func (s *monitoringServiceImpl) BatchRecordMetrics(ctx context.Context, metrics []struct {
	ServiceName string
	Type        int32
	Value       float64
	Unit        string
	Tags        map[string]string
}) error {
	var ms []*model.Metric
	now := time.Now().UnixMilli()
	for _, m := range metrics {
		ms = append(ms, &model.Metric{
			ID:          uuid.New().String(),
			ServiceName: m.ServiceName,
			Type:        int(m.Type),
			Value:       m.Value,
			Unit:        m.Unit,
			Tags:        dao.MarshalTags(m.Tags),
			Timestamp:   now,
		})
	}
	return s.repo.BatchSaveMetrics(ctx, ms)
}

func (s *monitoringServiceImpl) QueryMetrics(ctx context.Context, serviceName string, metricType int32, startTime, endTime int64, tags map[string]string) ([]*model.Metric, int64, error) {
	return s.repo.QueryMetrics(ctx, serviceName, int(metricType), startTime, endTime, tags, 1, 100)
}

func (s *monitoringServiceImpl) RecordLog(ctx context.Context, serviceName, level, message string, fields map[string]string) error {
	l := &model.LogEntry{
		ID:          uuid.New().String(),
		ServiceName: serviceName,
		Level:       level,
		Message:     message,
		Fields:      dao.MarshalTags(fields),
		Timestamp:   time.Now().UnixMilli(),
	}
	return s.repo.SaveLog(ctx, l)
}

func (s *monitoringServiceImpl) BatchRecordLogs(ctx context.Context, logs []struct {
	ServiceName string
	Level       string
	Message     string
	Fields      map[string]string
}) error {
	var ls []*model.LogEntry
	now := time.Now().UnixMilli()
	for _, l := range logs {
		ls = append(ls, &model.LogEntry{
			ID:          uuid.New().String(),
			ServiceName: l.ServiceName,
			Level:       l.Level,
			Message:     l.Message,
			Fields:      dao.MarshalTags(l.Fields),
			Timestamp:   now,
		})
	}
	return s.repo.BatchSaveLogs(ctx, ls)
}

func (s *monitoringServiceImpl) QueryLogs(ctx context.Context, serviceName, level, query string, startTime, endTime int64, page, pageSize int32) ([]*model.LogEntry, int64, error) {
	p, ps := int(page), int(pageSize)
	if p < 1 {
		p = 1
	}
	if ps < 1 || ps > 100 {
		ps = 20
	}
	return s.repo.QueryLogs(ctx, serviceName, level, query, startTime, endTime, p, ps)
}

func (s *monitoringServiceImpl) QueryAlerts(ctx context.Context, serviceName *string, level *int32, resolved *bool, startTime, endTime int64) ([]*model.Alert, int64, error) {
	var sn string
	if serviceName != nil {
		sn = *serviceName
	}
	var lvl *int
	if level != nil {
		l := int(*level)
		lvl = &l
	}
	return s.repo.QueryAlerts(ctx, sn, lvl, resolved, startTime, endTime)
}

func (s *monitoringServiceImpl) ResolveAlert(ctx context.Context, alertID string) error {
	return s.repo.ResolveAlert(ctx, alertID)
}

func (s *monitoringServiceImpl) UpdateServiceStatus(ctx context.Context, serviceName, status string, errorMessage *string, metadata map[string]string) error {
	ss := &model.ServiceStatus{
		ID:            uuid.New().String(),
		ServiceName:   serviceName,
		Status:        status,
		LastCheckTime: time.Now().UnixMilli(),
		ErrorMessage:  errorMessage,
		Metadata:      dao.MarshalTags(metadata),
	}
	return s.repo.UpsertServiceStatus(ctx, ss)
}

func (s *monitoringServiceImpl) GetServiceStatus(ctx context.Context, serviceName string) (*model.ServiceStatus, error) {
	return s.repo.GetServiceStatus(ctx, serviceName)
}

func (s *monitoringServiceImpl) ListServiceStatuses(ctx context.Context) ([]*model.ServiceStatus, error) {
	return s.repo.ListServiceStatuses(ctx)
}

func (s *monitoringServiceImpl) checkThreshold(ctx context.Context, m *model.Metric) {
	var threshold float64
	var level int
	var alertMsg string

	switch m.Type {
	case 6:
		threshold = 10.0
		level = 3
		alertMsg = fmt.Sprintf("错误率 %.2f%% 超过阈值 %.0f%%", m.Value, threshold)
	case 7:
		threshold = 5000.0
		level = 2
		alertMsg = fmt.Sprintf("延迟 %.2fms 超过阈值 %.0fms", m.Value, threshold)
	case 8:
		threshold = 100.0
		level = 2
		alertMsg = fmt.Sprintf("吞吐量 %.2f 低于阈值 %.0f", m.Value, threshold)
	default:
		return
	}

	shouldAlert := false
	switch m.Type {
	case 6, 7:
		shouldAlert = m.Value > threshold
	case 8:
		shouldAlert = m.Value < threshold
	}

	if shouldAlert {
		s.repo.CreateAlert(ctx, &model.Alert{
			ID:          uuid.New().String(),
			ServiceName: m.ServiceName,
			Level:       level,
			Message:     alertMsg,
			MetricValue: m.Value,
			Threshold:   threshold,
			Timestamp:   m.Timestamp,
		})
	}
}

File: web/src/api/monitoring.ts
import client from './client'

function extractData<T>(res: { data: unknown }, fallback: T): T {
  const d = res.data as Record<string, unknown>
  if (!d) return fallback
  if (d.data !== undefined && d.data !== null) return d.data as T
  if (typeof d === 'object' && !Array.isArray(d) && !('code' in d)) return d as T
  return fallback
}

export interface Metric {
  id?: string
  service_name: string
  type: number
  value: number
  unit?: string
  timestamp?: number
  tags?: Record<string, string>
}

export interface Alert {
  id?: string
  service_name: string
  level: number
  message: string
  metric_name: string
  metric_value: number
  threshold: number
  timestamp?: number
  resolved: boolean
  resolution_time?: string
}

export interface LogEntry {
  id?: string
  service_name: string
  level: string
  message: string
  timestamp?: number
  tags?: Record<string, string>
}

export interface ServiceStatus {
  service_name: string
  status: string
  last_check_time?: number
  error_message?: string
  metadata?: Record<string, string>
}

export async function recordMetric(metric: Omit<Metric, 'id' | 'timestamp'>) {
  const res = await client.post('/monitoring/metric', metric)
  return res.data
}

export async function batchRecordMetrics(metrics: Omit<Metric, 'id' | 'timestamp'>[]) {
  const res = await client.post('/monitoring/metric/batch', { metrics })
  return res.data
}

export async function queryMetrics(params: {
  service_name?: string
  type?: number
  start_time?: number
  end_time?: number
}): Promise<Metric[]> {
  try {
    const res = await client.get('/monitoring/metric', { params })
    return extractData<Metric[]>(res, [])
  } catch {
    return []
  }
}

export async function recordLog(log: Omit<LogEntry, 'id' | 'timestamp'>) {
  const res = await client.post('/monitoring/log', log)
  return res.data
}

export async function batchRecordLogs(logs: Omit<LogEntry, 'id' | 'timestamp'>[]) {
  const res = await client.post('/monitoring/log/batch', { logs })
  return res.data
}

export async function queryLogs(params: {
  service_name?: string
  level?: string
  query?: string
  start_time?: number
  end_time?: number
  page?: number
  page_size?: number
}): Promise<LogEntry[]> {
  try {
    const res = await client.get('/monitoring/log', { params })
    return extractData<LogEntry[]>(res, [])
  } catch {
    return []
  }
}

export async function queryAlerts(params: {
  service_name?: string
  level?: number
  resolved?: boolean
  start_time?: number
  end_time?: number
}): Promise<Alert[]> {
  try {
    const res = await client.get('/monitoring/alert', { params })
    return extractData<Alert[]>(res, [])
  } catch {
    return []
  }
}

export async function resolveAlert(alertId: string) {
  const res = await client.put(`/monitoring/alert/${alertId}/resolve`)
  return res.data
}

export async function updateServiceStatus(status: {
  service_name: string
  status: string
  error_message?: string
  metadata?: Record<string, string>
}) {
  const res = await client.put('/monitoring/service-status', status)
  return res.data
}

export async function getServiceStatus(serviceName: string): Promise<ServiceStatus | null> {
  try {
    const res = await client.get('/monitoring/service-status', { params: { service_name: serviceName } })
    return extractData<ServiceStatus | null>(res, null)
  } catch {
    return null
  }
}

export interface ServiceInfo {
  name: string
  port: number
  address: string
  etcd_name: string
}

export async function listServices(): Promise<ServiceInfo[]> {
  try {
    const res = await client.get('/monitoring/services')
    return extractData<ServiceInfo[]>(res, [])
  } catch {
    return []
  }
}

export async function listServiceStatus(): Promise<ServiceStatus[]> {
  try {
    const res = await client.get('/monitoring/service-status/list')
    return extractData<ServiceStatus[]>(res, [])
  } catch {
    return []
  }
}

export interface BenchConfig {
  url: string
  users: number
  duration: number
  rps: number
  scenario: string
}

export interface BenchResult {
  total: number
  success: number
  failed: number
  timeout: number
  successRate: number
  qps: number
  avgLatency: number
  minLatency: number
  maxLatency: number
  p50: number
  p90: number
  p95: number
  p99: number
  statusCodes: Record<number, number>
  errors: Record<string, number>
  running: boolean
  elapsed: number
}

export const BENCH_SCENARIOS = [
  { value: 'health', label: '健康检查' },
  { value: 'login', label: '用户登录' },
  { value: 'get_user', label: '获取用户' },
  { value: 'chat_history', label: '聊天历史' },
  { value: 'send_message', label: '发送消息' },
  { value: 'bot_list', label: 'Bot列表' },
  { value: 'contacts', label: '联系人' },
  { value: 'mixed', label: '混合场景' },
]
